Review work on Feature/text analytics language api

maririos opened the pull request:
The purpose of this PR is to review the progress done in the feature/text-analytics-language-api branch with no intentions of merging the branch to master.

nisha-bhatia committed:
rebase with main

nisha-bhatia committed:
rebase with main

nisha-bhatia committed:
wip

nisha-bhatia committed:
wip

nisha-bhatia committed:
wip

maririos commented on sdk/textanalytics/Azure.AI.TextAnalytics/src/Transforms.cs:
```diff
@@ -32,6 +32,22 @@ internal static TextAnalyticsError ConvertToError(TextAnalyticsErrorInternal err
             return new TextAnalyticsError(errorCode, message, target);
         }
 
+        internal static TextAnalyticsError ConvertToError(DocumentError error)
```
From this code, it looks like you are only using the `error.Error` property, so instead of passing the whole `DocumentError` object, consider sending to this function just `DocumentError.Error`.

Context: `DocumentError` has 2 main properties: the `id` of the document and the `error` itself. This function is already called per each document, which is why you never use `id` in here.

nisha-bhatia replied:
updated

maririos commented on sdk/textanalytics/Azure.AI.TextAnalytics/src/Transforms.cs:
```diff
@@ -73,6 +89,17 @@ internal static DetectedLanguage ConvertToDetectedLanguage(DocumentLanguage docu
+        internal static DetectedLanguage ConvertToDetectedLanguage(LanguageDetectionDocumentResult documentLanguage)
+        {
+            List<TextAnalyticsWarning> warnings = new List<TextAnalyticsWarning>();
+            foreach (var warning in documentLanguage.Warnings)
```
If the service doesn't return `Warnings`, `document.Language.Warnings` is going to be null and the code here is going to throw.

Because this check and loop happen in every endpoint, we extracted that logic into the [ConvertToWarnings](https://github.com/Azure/azure-sdk-for-net/blob/2c9bf3756510134d789bf86e85c3ce7ab013b6d5/sdk/textanalytics/Azure.AI.TextAnalytics/src/Transforms.cs#L67) method.
We should do the same or similar for new code too.

nisha-bhatia replied:
added a ConvertToWarnings method

maririos commented on sdk/textanalytics/Azure.AI.TextAnalytics/src/Generated/Models/DetectedLanguageInternal.cs:
```diff
@@ -14,7 +14,7 @@ namespace Azure.AI.TextAnalytics.Models
         /// <param name="iso6391Name"> A two letter representation of the detected language according to the ISO 639-1 standard (e.g. en, fr). </param>
         /// <param name="confidenceScore"> A confidence score between 0 and 1. Scores close to 1 indicate 100% certainty that the identified language is true. </param>
-        internal DetectedLanguageInternal(string name, string iso6391Name, double confidenceScore)
+        public DetectedLanguageInternal(string name, string iso6391Name, double confidenceScore)
```
curious, why the need to make this class public? also, did you make the class `public` directly here? or in another file that I missed?

nisha-bhatia replied:
I think this should be reverted

maririos commented on sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs:
```diff
@@ -19,7 +20,8 @@ namespace Azure.AI.TextAnalytics
     {
         private readonly Uri _baseUri;
-        internal readonly TextAnalyticsRestClient _serviceRestClient;
+        internal readonly AnalyzeTextRestClient _serviceRestClient;
+        internal readonly MicrosoftCognitiveLanguageServiceRestClient _cognitiveRestClient;
```
what is the difference between these two? I don't see the `MicrosoftCognitiveLanguageServiceRestClient` referenced here

maririos commented on sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs:
```diff
@@ -65,7 +67,7 @@ public TextAnalyticsClient(Uri endpoint, TokenCredential credential, TextAnalyti
 
             var pipeline = HttpPipelineBuilder.Build(options, new BearerTokenAuthenticationPolicy(credential, defaultScope));
-            _serviceRestClient = new TextAnalyticsRestClient(_clientDiagnostics, pipeline, endpoint.AbsoluteUri, TextAnalyticsClientOptions.GetVersionString(options.Version));
+            _serviceRestClient = new AnalyzeTextRestClient(_clientDiagnostics, pipeline, endpoint.AbsoluteUri, TextAnalyticsClientOptions.GetVersionString(options.Version));
```
we need to instantiate `_cognitiveRestClient` too otherwise the calls are not going to get executed

maririos replied:
actually, I am unclear what the difference is between the two. Could you explain when to use which?

maririos commented on sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs:
```diff
@@ -195,17 +203,26 @@ public virtual Response<DetectedLanguage> DetectLanguage(string document, string
 
-                if (result.Value.Errors.Count > 0)
+                var languageDetection = result.Value as LanguageDetectionTaskResult;
+                Debug.Assert(languageDetection != null);
+                Response response = result.GetRawResponse();
```
need to cleanup these 2 lines

maririos commented on sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs:
```diff
@@ -195,17 +203,26 @@ public virtual Response<DetectedLanguage> DetectLanguage(string document, string
+                Response<AnalyzeTextTaskResult> result = _cognitiveRestClient.AnalyzeText(analyzeLanguageDetection, cancellationToken: cancellationToken);
 
-                if (result.Value.Errors.Count > 0)
+                var languageDetection = result.Value as LanguageDetectionTaskResult;
```
before you were doing a [cast expression](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast#cast-expression) `var languageDetection = (LanguageDetectionTaskResult)result.Value;` and here you are using the `as` [operator](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast#as-operator).

which do you think should be the patter we should use everywhere?

maririos commented on sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs:
```diff
@@ -364,6 +377,33 @@ private async Task<Response<DetectLanguageResultCollection>> DetectLanguageBatch
+            try
+            {
+                var analyzeLanguageDetection = new AnalyzeTextLanguageDetectionInput { AnalysisInput = batchInput };
+                analyzeLanguageDetection.Parameters.ModelVersion = options.ModelVersion;
+                analyzeLanguageDetection.Parameters.LoggingOptOut = options.DisableServiceLogs;
```
curious why the `Parameters` are treated differently than `AnalysisInput`

nisha-bhatia replied:
I updated the Parameter construction to match the AnalysisInput

maririos commented on sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs:
```diff
@@ -195,17 +203,26 @@ public virtual Response<DetectedLanguage> DetectLanguage(string document, string
 
-                if (result.Value.Errors.Count > 0)
+                var languageDetection = result.Value as LanguageDetectionTaskResult;
+                Debug.Assert(languageDetection != null);
```
need to clean up this line

maririos commented on sdk/textanalytics/Azure.AI.TextAnalytics/src/Azure.AI.TextAnalytics.csproj:
```diff
@@ -24,4 +24,8 @@
   </ItemGroup>
+
+  <ItemGroup>
+    <Folder Include="Models\" />
```
curious what this is for?

nisha-bhatia replied:
I'm not sure how this got updated

maririos replied:
Jesse fixed this. resolving comment

azure-sdk commented:
**API change check for `Azure.AI.TextAnalytics`**

API changes have been detected in `Azure.AI.TextAnalytics`. You can review API changes [here](https://apiview.dev/Assemblies/Review/fffbbfca48b8449a89b5a5a1d6209b75?diffOnly=True&diffRevisionId=f08a286d3f5a43ad9f1c16cad078b5af)

maririos commented on sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs:
```diff
@@ -241,11 +258,12 @@ public virtual Response<DetectedLanguage> DetectLanguage(string document, string
-            LanguageBatchInput detectLanguageInputs = ConvertToLanguageInputs(documents, countryHint);
-
-            return await DetectLanguageBatchAsync(detectLanguageInputs, options, cancellationToken).ConfigureAwait(false);
+            //Argument.AssertNotNullOrEmpty(documents, nameof(documents));
```
curious, why did you have to comment these lines out? I thought this path was already working

maririos commented on sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs:
```diff
@@ -637,31 +674,37 @@ public virtual async Task<Response<RecognizeEntitiesResultCollection>> Recognize
-                    cancellationToken).ConfigureAwait(false);
-                var response = result.GetRawResponse();
+                var input = entityRecognitionInput.AnalysisInput;
+                var analyzeRecognizeEntities = new AnalyzeTextEntityRecognitionInput { AnalysisInput = input };
```
could we eliminate `input` and pass the value directly?

maririos commented on sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs:
```diff
@@ -429,21 +449,26 @@ public virtual async Task<Response<CategorizedEntityCollection>> RecognizeEntiti
-
-                Response<EntitiesResult> result = await _serviceRestClient.EntitiesRecognitionGeneralAsync(
-                    new MultiLanguageBatchInput(documents),
-                    stringIndexType: Constants.DefaultStringIndexType,
```
this parameter is no longer passed. Could you please add it again?

maririos replied:
I see this one is missing in other places too. 
`StringIndexType` is what tells the service what type of encoding they should use when managing strings. (think for example when someone sends an emoji. According to the encoding, the properties of a string will change).
By default, .NET manages `Utf16codeUnit` so we always need to send this property to the service. See how the library has a default value [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/textanalytics/Azure.AI.TextAnalytics/src/Constants.cs#L10).

maririos commented on sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs:
```diff
@@ -744,26 +791,26 @@ public virtual async Task<Response<PiiEntityCollection>> RecognizePiiEntitiesAsy
+                Response<AnalyzeTextTaskResult> result = await _cognitiveRestClient.AnalyzeTextAsync(analyzePiiEntites, cancellationToken: cancellationToken).ConfigureAwait(false);
 
-                Response<PiiEntitiesResult> result = await _serviceRestClient.EntitiesRecognitionPiiAsync(
-                    new MultiLanguageBatchInput(documents),
```
We need to make sure we send these properties too

maririos commented on sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsOperationStatus.cs:
```diff
@@ -11,5 +11,8 @@ namespace Azure.AI.TextAnalytics
     {
+        /// <summary> Added for compilation. </summary>
+        /// TODO: delete later
+        public static TextAnalyticsOperationStatus Rejected { get; } = new TextAnalyticsOperationStatus(CancellingValue);
```
could u add an issue for this so we don't forget to clean it up? it needs to happen during this release cycle

nisha-bhatia replied:
Issue: https://github.com/Azure/azure-sdk-for-net/issues/28326

nisha-bhatia replied:
https://github.com/Azure/azure-sdk-for-net/issues/28326

maririos commented on sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClientDiagnostics.cs:
```diff
@@ -32,28 +33,29 @@ public TextAnalyticsClientDiagnostics(ClientOptions options) : base(options)
             ResponseHeaders responseHeaders,
             ref IDictionary<string, string>? additionalInfo)
         {
-            if (!string.IsNullOrEmpty(content))
```
curious about why was this causing a problem? 
can you create an issue to undo this?

maririos replied:
Fixed in PR https://github.com/Azure/azure-sdk-for-net/pull/28449

maririos committed:
Parity with v3.2 (#28290)

maririos committed:
update swagger (#28335)

maririos committed:
[TA] Merge rest clients into one by renaming OperationIds (#28340)

maririos committed:
add new version, constant version for mocks and samples (#28346)

jsquire committed:
[Text Analytics] Legacy API and Abstractions (#28229)

The focus of these changes is add the generated surface for the the
legacy Text Analytics REST API in a stand-alone project that is compatible
with the AutoRest generator and its conventions.  The types for the legacy API
are generated under a unique namespace and will be included in the Text Analytics
client library via source linking.  This is intended to allow the legacy API
to be managed and regenerated in a standard way while avoiding conflicts
with the Cognitive Language API that will be used for service operations on
new version going forward.

Also included are the stubs for the `ServiceClient` abstractions, forming
the basis of the contract by which the `TextAnalyticsClient` will be able to
target different REST services through a common interface.

nisha-bhatia committed:
Add LanguageDetection Endpoint for new convenience layer (#28359)

jsquire committed:
[Text Analytics] Legacy Service Client (#28423)

* [Text Analytics] Legacy Service Client

The focus of these changes is to introduce the service client responsible
for interacting with the legacy Text Analytics REST API.  This client
accepts and returns types from the SDK's public API, translating them as
needed for service operations.

maririos committed:
[TA] Test enabled for Detect Language (#28449)

nisha-bhatia committed:
Add RecognizeEntities to Azure.AI.TextAnalytics (#28433)

jsquire committed:
[Text Analytics] Delegation to Serivce Clients (#28463)

* [Text Analytics] Delegation to Serivce Clients

The focus of these changes is to shift functionality for the Language REST
API integration into the `LanguageServiceClient` and update the `TextAnalyticsClient`
to delegate all operations to the service client.

nisha-bhatia committed:
Add PiiEntities to Azure.AI.TextAnalytics (#28470)

maririos committed:
[TA] Reuse ConvertToMultiLanguageInputs functions (#28473)

maririos committed:
[TA] Add test for recognize entities (#28474)

jsquire committed:
[Text Analytics] Fixing CodeGen Capitalization (#28512)

The focus of these changes is to fix a typo in the
`CustomEntitiesResultDocumentsItem` stub which was causing "item" to be
generated as lower-case. This differs from the capitalization of the file
committed to the repository and was causing issues with case-sensitive
file systems in CI when the "Analyze" step was run during API diff checks.

maririos committed:
fix, live tests, samples (#28484)

nisha-bhatia committed:
Add LinkedEntities to Azure.AI.TextAnalytics (#28494)

maririos committed:
non-batch endpoints working and disable lro samples (#28511)

nisha-bhatia committed:
Update LanguageServiceClient.cs (#28522)

maririos committed:
Keyphrases with tests (#28524)

maririos committed:
enable tests (#28526)
